src/query/api/v1/handler/prometheus/debugger/validate.go

```diff
@@ -0,0 +1,266 @@
+func NewPromDebugHandler(
+	h *native.PromReadHandler,
+	scope tally.Scope,
+) *PromDebugHandler {
+	return &PromDebugHandler{
+		scope:       scope,
+		readHandler: h,
+	}
+}
+
+func (h *PromDebugHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx := context.WithValue(r.Context(), handler.HeaderKey, r.Header)
+	logger := logging.WithContext(ctx)
+
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		logger.Error("unable to read data", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	var promDebug prometheus.PromDebug
+	if err := json.Unmarshal(data, &promDebug); err != nil {
+		logger.Error("unable to unmarshal data", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	s, err := debug.NewStorage(promDebug.Input)
+	if err != nil {
+		logger.Error("unable to create storage", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	h.engine = executor.NewEngine(s, h.scope.SubScope("debug_engine"))
+	results, _, err := h.readHandler.ServeHTTPWithEngine(w, r, h.engine)
+	if err != nil {
+		logger.Error("unable to read data", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	promResults, err := debug.PromResultToSeriesList(promDebug.Results, models.NewTagOptions())
+	if err != nil {
+		logger.Error("unable to convert prom results data", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	mismatches, err := validate(tsListToMap(promResults), tsListToMap(results))
+	if err != nil {
+		logger.Error("error validating results", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+
+	if err := renderDebugMismatchResultsJSON(w, mismatches); err != nil {
+		logger.Error("unable to write back mismatch data", zap.Error(err))
+		xhttp.Error(w, err, http.StatusBadRequest)
+		return
+	}
+}
+
+func tsListToMap(tsList []*ts.Series) map[string]*ts.Series {
+	tsMap := make(map[string]*ts.Series, len(tsList))
+
+	for _, series := range tsList {
+		series.Tags = series.Tags.Normalize()
+		id := series.Tags.ID()
+		tsMap[id] = series
+	}
+
+	return tsMap
+}
+
+type mismatch struct {
+	seriesName       string
+	promVal, m3Val   float64
+	promTime, m3Time time.Time
+}
+
+func validate(prom, m3 map[string]*ts.Series) ([][]mismatch, error) {
+	if len(prom) != len(m3) {
+		return [][]mismatch{}, errors.New("number of Prom series not equal to number of M3 series")
+	}
+
+	var mismatches [][]mismatch
+
+	for id, promSeries := range prom {
+		var (
+			m3Series *ts.Series
+			exists   bool
+		)
+
+		if m3Series, exists = m3[id]; !exists {
+			return [][]mismatch{}, fmt.Errorf("series with id %s does not exist in M3 results", id)
+		}
+
+		promdps := promSeries.Values().Datapoints()
+		m3dps := m3Series.Values().Datapoints()
+
+		var mismatchList []mismatch
```

Comment:
initialize this with a length

Reply:
Same comment as above, we don't know the length upfront.